### PR TITLE
Fix issues with ordering of prompts in fairseq-generate.

### PR DESCRIPTION
**Pull request by mayhewsw**
Note: this merely fixes a bug in the code stubs provided for fairseq models. Neither the data nor the scoring functions have been changed.

#3 pointed out that generating prompts from text wasn't always guaranteed to work because of processing that happened in the fairseq pipeline (whitespace in particular). This PR assumes that the output is in the same order as the original prompt file and matches original prompt IDs to the output prompts.

One note: fairseq-generate does not generate output in the order that it receives it, although it does retain the ordering in the metadata. I had to jump through a couple hoops to solve this.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+data/
+checkpoints/
+.vscode
```
Comment (cpbrust) [inline .gitignore]: Nit: newline

File: my_cands_extract.py [hunk 2]
```diff
@@ -1,60 +1,93 @@
...
+        sline = line.strip().split("\t")
+        if line.startswith("S-"):
+            num = int(sline[0].split("-")[-1])
+            outd[num].append(sline[1])
+            first = True
+            cands = 0
+        elif line.startswith("T-"):
+            num = int(sline[0].split("-")[-1])
+            # this is the reference
+            #f.write(sline[1] + "\n")
```
Comment (cpbrust) [inline my_cands_extract.py]: Extra comment

File: my_cands_extract.py [hunk 3]
```diff
@@ -1,60 +1,93 @@
...
+            refd[num].append(sline[1] + "\n")
+        elif line.startswith("H-"):
+            num = int(sline[0].split("-")[-1])
+            # this is the prediction, there may be many of these.
+            if candlimit == -1 or cands < candlimit:
+                #out.write(sline[2] + "\n")
```
Comment (cpbrust) [inline my_cands_extract.py]: Extra comment

File: my_cands_extract.py [hunk 4]
```diff
@@ -1,60 +1,93 @@
...
+                outd[num].append(sline[2] + "\n")
+                cands += 1
+
+            # only write the first of these.
+            if first:
+                #sf.write(sline[2] + "\n")
```
Comment (cpbrust) [inline my_cands_extract.py]: Extra comment

File: preprocess.sh [hunk 5]
```diff
@@ -70,4 +70,6 @@ mkdir ${DATA}/bin
 # Since we only provide training data, you may want to make your own split.
 fairseq-preprocess --source-lang $src --target-lang $tgt \
     --trainpref $DATA/train-${outf} --validpref $DATA/dev-${outf} --testpref $DATA/test-${outf} \
-    --destdir ${DATA}/bin/ --workers 20 --dataset-impl raw
+    --destdir ${DATA}/bin/ --workers 20 --dataset-impl raw \
+    # --srcdict checkpoints/${src}_${tgt}/dict.${src}.txt \
```
Comment (cpbrust) [inline preprocess.sh:71]: Extra comments

File: run_pretrained.sh [hunk 6]
```diff
@@ -16,15 +16,18 @@ fairseq-generate data/courses/en-${tgt}/bin/ --path $MODEL --dataset-impl raw --
    --beam $NBEST --batch-size 128 --remove-bpe --nbest $NBEST --replace-unk > $OUTPUT/gen.out
  
 # this cleans all the BPE
-sed -i 's/@@ //g' $OUTPUT/gen.out
+sed -i '' 's/@@ //g' $OUTPUT/gen.out
 
-python my_cands_extract.py --infile $OUTPUT/gen.out --outfile $OUTPUT/all_cands.txt --candlimit $CANDLIMIT
+TESTFILE=${SHARED_TASK_DATA}/${src}_${tgt}/test.${src}_${tgt}.2020-01-13.gold.txt
```
Comment (cpbrust) [inline run_pretrained.sh:21]: Did you mean to keep this?

File: variables.sh [hunk 7]
```diff
@@ -12,6 +12,9 @@ DUMP=data/dumps/${lang}
 MOSES=$HOME/IdeaProjects/mosesdecoder
 SUBWORDNMT=$HOME/IdeaProjects/subword-nmt
 
+# Location of the shared task data.
+SHARED_TASK_DATA=$HOME/staple2020/staple-2020-train
```
Comment (cpbrust) [inline variables.sh:16]: Is this used? It looks duplicated against a variable in `preprocess.sh`

**Review by cpbrust — CHANGES_REQUESTED** [review]

File: preprocess.sh (hunk 5 above)
Comment (mayhewsw) [reply preprocess.sh:71]: I actually want this comment here, as it could be useful.

File: run_pretrained.sh (hunk 6 above)
Comment (mayhewsw) [reply run_pretrained.sh:21]: Yes.

File: variables.sh (hunk 7 above)
Comment (mayhewsw) [reply variables.sh:16]: Ah, thanks. It shouldn't be in `preprocess.sh`

**Review by cpbrust — APPROVED** [review]

**Pull request merged**